### PR TITLE
[RDY][CR] Mark the shortest route to a refugee center on map

### PR DESCRIPTION
**Pull request by codemime**
This partly covers #15979 (milestone, no mission yet) and implements a [suggestion](https://github.com/CleverRaven/Cataclysm-DDA/issues/15979#issuecomment-203228520) of @Coolthulhu .
### Observable behavior

As it's said in the title: the shortest route (if there's one) gets revealed along with the center itself.
### Implementation details
- The A\* pathfinding algorithm was already implemented in the `overmap` class, but was badly intermixed with a routine that adds roads (`overmap::make_hiway()`). Now it's extracted to `find_path()` and is free of too specific code (a lambda that estimates nodes for the algorithm is accepted)
- The new `find_path()` was reused for, this time, finding and revealing roads
- The monstrous `computer::activate_function()` is now a wee bit shorter
- `tripoint`s were used here and there instead of plain coordinates. The code, however, needs some further refactoring.

Also:
- Eliminated some duplication in overmap bounds checking (via introducing `overmap::inbounds()` with a `clearance` param).
- Removed one unused function (`overmap::is_road_or_highway()`).

EDIT:
- Pathfinding for the `map` class was renamed to reflect its purpose (and its owner) more clearly
- The A\* algorithm is now separated from `overmap` and transformed into the general-purpose template
### Known limitations

~~Pathfinding works only locally (within one overmap) and the route won't be marked if the refugee center is any farther (in which case everything works like before). WIP because I'd try to overcome the limitation.~~

Inter-overmap pathfinding is now possible (the area is limited up to 8x8 overmaps, but can be increased if needed for other purposes).

**Route to a distant center:**

refu cen

### CODE REVIEW

**Comment by Lady-Luck** [comment]
This looks really cool.

**Comment by codemime** [comment]
Thank you.

**Comment by Coolthulhu** [comment]
I don't like that pathfinding rename. Pathfinding on local map is the default pathfinding as it is used the most.

Generic pathfinding will most likely be slow. Especially since you provide it with a generic function object that isn't guaranteed to be inlined.
It would be safer to just specialize it as overmap pathfinding. Since you don't allow diagonals, it is not generic pathfinding anyway.

For some reason overmap access tends to be stupidly slow already. At least when reading maps. Could be because it tries to generate other overmaps, though.

**Comment by codemime** [comment]
> I don't like that pathfinding rename. Pathfinding on local map is the default pathfinding as it is used the most.

No problem, will rename it back. 

> It would be safer to just specialize it as overmap pathfinding. Since you don't allow diagonals, it is not generic pathfinding anyway.

Agreed, it's not, but I'd like to keep it independent.

> For some reason overmap access tends to be stupidly slow already. At least when reading maps. Could be because it tries to generate other overmaps, though.

Yes, overmap generating is one reason, but the most annoying thing is overmap terrain comparison: it's done by comparing strings (the stupid way). Simply finding a refugee center takes about 1.3 seconds on my (quite fast) computer. I'm going to fix this in a separate PR later on.

**Comment by codemime** [comment]
Ready to be reviewed/tested.

**Comment by Coolthulhu** [comment]
Got a segfault.

**Comment by Coolthulhu** [comment]
Can't find the reason in the code. That `nodes[i]` is hard to read, would be better to extract them into two named references.

**Comment by codemime** [comment]
> Got a segfault.

How exactly?

**Comment by Coolthulhu** [comment]
Went up to a console and picked "contact us". The game froze and after a while segfaulted.

**Comment by codemime** [comment]
I'm trying and failing to reproduce. Can it be stack-size related (the arrays are eight times bigger in case of global pathfinding)? Maybe switch to dynamically allocated memory?

**Comment by Coolthulhu** [comment]
Could be it.
Alternatively bounds checking could be too forgiving. Did you test on a map with no valid target?

**Comment by codemime** [comment]
No, I didn't (deliberately at least), but the target was farther than two overmaps a couple of times (and no route was found). Disabling the target completely is a good idea.

**Comment by codemime** [comment]
Large arrays are now allocated on the heap (I didn't see any performance loss) and possible sources of range check errors are eliminated. Should be stable now.

**Comment by codemime** [comment]
Rebased.

File: src/simple_pathfinding.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    bool operator< ( const node &n ) const {
+        return priority > n.priority;
+    }
+};
+
+/// @param estimate BinaryPredicate( node &previous, node &current ) returns
+/// integer estimation (smaller - better) for the current node or a negative value
+/// if the node is unsuitable.
+template<class BinaryPredicate>
+std::vector<node> find_path( const tripoint &source,
```
Comment (mugling) [inline src/simple_pathfinding.h]: The plugin in estimator is a good design choice

File: src/simple_pathfinding.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+{
+
+struct node {
+    int x;
+    int y;
+    int dir;
+    int priority;
+
+    node( int x, int y, int dir, int priority = 0 ) :
+        x( x ),
```
Comment (mugling) [inline src/simple_pathfinding.h]: Preferably on one line, especially for trivial structs with few arguments. Not critical and worth delaying this PR on but worth considering in future.

File: src/simple_pathfinding.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
+#ifndef PATHFINDIND_H
+#define PATHFINDIND_H
+
+#include "debug.h"
+#include "enums.h"
+
+#include <queue>
+#include <vector>
+
+namespace pf
```
Comment (mugling) [inline src/simple_pathfinding.h]: Somewhat ambivalent about the namespace, could be more descriptive

File: src/overmap.cpp [hunk 4]
```diff
@@ -3406,133 +3417,44 @@ void overmap::place_rifts(int const z)
...
+            return -1;
+        }
+
+        int res = ( std::abs( dest.x - cur.x ) + std::abs( dest.y - cur.y ) ) / disp;
+        // Prefer existing roads.
+        res += check_ot_type( base, cur.x, cur.y, dest.z ) ? 0 : 3;
+        // Prefer flat land over bridges
+        res += !is_river( ter( cur.x, cur.y, dest.z ) ) ? 0 : 2;
+        // Try not to turn too much
+        //res += (mn.d == d) ? 0 : 1;
```
Comment (mugling) [inline src/overmap.cpp]: Can you elaborate?

**Comment by mugling** [comment]
Well though out. Hopefully some more mapgen PR's will folllow

File: src/simple_pathfinding.h (hunk 1 above)
Comment (codemime) [inline src/simple_pathfinding.h]: Thanks. I was inspired by `std` functions.

File: src/overmap.cpp (hunk 4 above)
Comment (codemime) [inline src/overmap.cpp]: For some reason this line was commented out in `overmap::make_hiway()` and I kept it intact.
It's supposed to slightly penalize direction changes to make roads less curvy. What shall we do?

**Comment by codemime** [comment]
Thank you. They certainly will.

**Pull request merged**